src/corona_hakab_model/infection.py

```diff
@@ -121,9 +155,20 @@ def _get_random_infection_info(self, agent_id: int, infection_probs: np.ndarray)
         connection_type = np.random.choice(len(infection_probs), p=infection_probs / np.sum(infection_probs))
 
         # determine infector
-        infectious_circle = \
-            [c for c in self.manager.social_circles_by_agent_index[agent_id] if c.connection_type == connection_type][0]
-        agents_id = [a.index for a in infectious_circle.agents]
+        agents_id = []
+
+        if connection_type in connection_types.With_Random_Connections:
+            infectious_circle = next(circle for circle in self.manager.social_circles_by_agent_index[agent_id]
+                                     if circle.connection_type == connection_type)
+            agents_id = [agent.index for agent in infectious_circle.agents]
+
+        if connection_type in connection_types.With_Geo_Random_Connections:
```

Comment:
There are (and probably will always be) _O(1)_ geo-random connection type and _O(1)_ geo-circles, hence _O(1)_ possible lists of agents_id.
However, this list is constructed again and again each time this function is called and the connection_type=WORK is chosen. 

Maybe you should store a dictionary of [all the people with circle type Y in geo-circle X] to allow quick access to this data. (There are many such dictionaries in the manager \ circ generator)